sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/epkversion/FeedRangeThroughputControlConfigManager.java

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation.changefeed.epkversion;
+
+import com.azure.cosmos.ThroughputControlGroupConfig;
+import com.azure.cosmos.ThroughputControlGroupConfigBuilder;
+import com.azure.cosmos.implementation.changefeed.ChangeFeedContextClient;
+import com.azure.cosmos.implementation.changefeed.Lease;
+import com.azure.cosmos.implementation.feedranges.FeedRangeEpkImpl;
+import com.azure.cosmos.implementation.routing.PartitionKeyInternalHelper;
+import com.azure.cosmos.models.FeedRange;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Mono;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+
+// Only used in CFP when customer configure throughput control config
+// The main purpose of this class is to monitor the lease feedRange -> partitionKeyRange mapping changes, and create corresponding throughput control group
+public class FeedRangeThroughputControlConfigManager {
+    private static final Logger logger = LoggerFactory.getLogger(FeedRangeThroughputControlConfigManager.class);
+
+    private final ThroughputControlGroupConfig throughputControlGroupConfig;
+    private final ChangeFeedContextClient documentClient;
+    private final AtomicReference<List<FeedRangeEpkImpl>> leaseTokens; // epk leases
+
+    public FeedRangeThroughputControlConfigManager(
+        ThroughputControlGroupConfig throughputControlGroupConfig,
+        ChangeFeedContextClient documentClient) {
+
+        checkNotNull(throughputControlGroupConfig, "Argument 'throughputControlGroupConfig' can not be null");
+        checkNotNull(documentClient, "Argument 'documentClient' can not be null");
+
+        this.throughputControlGroupConfig = throughputControlGroupConfig;
+        this.documentClient = documentClient;
+        this.leaseTokens = new AtomicReference<>();
+    }
+
+    /**
+     * This method will be called during the leases load balancing time from {@link com.azure.cosmos.implementation.changefeed.common.EqualPartitionsBalancingStrategy}.
+     * We are going to track the up to date all leases and refresh the partitionKeyRangesCache
+     *
+     * @param leases all the current leases in lease container.
+     * @return a representation of the deferred computation of this call.
+     */
+    public Mono<Void> refresh(List<Lease> leases) {
+        if (leases != null && !leases.isEmpty()) {
+            this.leaseTokens.set(leases.stream().map(lease -> (FeedRangeEpkImpl)lease.getFeedRange()).collect(Collectors.toList()));
+        }
+
+        return this.documentClient.getOverlappingRanges(PartitionKeyInternalHelper.FullRange, true)
```

Comment:
As discussed offline - force refresh of PKRange/FeedRange is not a good idea every 13 seconds (or any cx configurable interval). Let's make this a random interval of at least 30 minutes or so.

Reply:
You are right Fabian, I do not need to to forceRefresh here. If a lease is being assigned to a CFP with staled cache, the ReadFeed requests will fail which eventually will trigger a forceRefresh

the logic has been changed into following:
In the `FeedRangeThroughputControlConfigManager`, maintain the following two maps:
    pkRangesToFeedRange
    feedRangeToThroughputControlGroupConfig

when refresh() being called, which usually happens during the lease load balancing stage, it will check whether the pkRange still exists in the current cache, if not, then we know some change has happened for the pkRange, its presence will be removed from the map, and then all the mapped feedRanges throughputControlGroupConfig will be removed from the feedRangeToThroughputControlGroupConfig map. 

when the feedRange being empty, it will trigger a re-recreation of the config so the partitionDivideFactor will be re-calcualted